### ISSUE TITLE
nextflow parent process hangs indefinitely at "Finishing pending tasks before exit" when using "finish" errorStrategy

### ISSUE THREAD

**Issue by ahstram** [nextflow-io/nextflow#3749]
Hi there,

I've noticed that Nextflow may hang if a process errors out, but only when using the  "finish" error strategy. If I switch back to the "terminate" error strategy, the parent process will exit out as expected.

My apologies for the rather strange looking example-- I struggled a bit to make an example from scratch, so I had to derive this from an example error we had in a more complicated code base.

Here's what my dummy `main.nf` looks like:

```groovy
#!/usr/bin/env nextflow

process failer {
    input:
        val(str)
    output:
        path("output.txt")
    shell:
    '''
        false
    '''
}

process passer {
    input:
        val(str)
    output:
        path("output.txt")
    shell:
    '''
        echo "!{str}" > output.txt
    '''
}

workflow {
    ch = Channel.from('foo','bar')
    fails = failer(ch).flatten()
    passes = passer('foobar')
    list = passes.concat(fails)
}
```

and here's my `nextflow.config`:

```groovy
process {
    errorStrategy= 'finish'
}
```

With the above `nextflow.config`, the nextflow process will write this to STDOUT/STDERR, **then hang indefinitely**:

```
$ nextflow run -ansi-log false main.nf
N E X T F L O W  ~  version 23.02.1-edge
Launching `main.nf` [hungry_plateau] DSL2 - revision: 4afb9ef9f6
[2b/1ddf67] Submitted process > failer (2)
[8f/b19a4f] Submitted process > failer (1)
[4e/e2115d] Submitted process > passer
ERROR ~ Error executing process > 'failer (2)'

Caused by:
  Process `failer (2)` terminated with an error exit status (1)

Command executed:

  false

Command exit status:
  1

Command output:
  (empty)

Work dir:
  /home/astram/work/2b/1ddf6708388b9fc1a6613f310af0d3

Tip: when you have fixed the problem you can continue the execution adding the option `-resume` to the run command line

 -- Check '.nextflow.log' file for details
Execution cancelled -- Finishing pending tasks before exit
```

and the last line of `.nextflow.log` is this:

```
Mar-13 13:54:39.777 [main] DEBUG nextflow.Session - Session await > all barriers passed
```

However, if I simply change the above `nextflow.config` file to:

```groovy
process {
    errorStrategy= 'terminate'
}
```

The parent nextflow process will terminate immediately.

Any ideas on what may be causing this? Seems like it is specific to the "finish" errorStrategy logic.

Thanks,

Alex

**Comment by pditommaso** [comment]
This is an interesting puzzle. First of all, it happens in a non-deterministic manner. Then, for some reason, it looks related to the use of `.flatten()`, removing it the issue does not arise any more.

**Comment by pditommaso** [comment]
Solved by d99b3432.
